Buildings/Controls/OBC/Utilities/PIDWithAutotuning/FirstOrderAMIGO.mo

```diff
@@ -0,0 +1,555 @@
+  connect(conProMod.tunSta, samk.trigger) annotation (Line(points={{62,62},{62,
+          6},{-90,6},{-90,-8}}, color={255,0,255}));
+  connect(samTi.trigger, conProMod.tunSta) annotation (Line(points={{-110,-38},
+          {-108,-38},{-108,6},{62,6},{62,62}}, color={255,0,255}));
+  connect(samTd.trigger, conProMod.tunSta) annotation (Line(points={{-130,-68},
+          {-130,6},{62,6},{62,62}}, color={255,0,255}));
+  connect(sam_u_s.trigger, triTun) annotation (Line(points={{-90,242},{-90,252},
+          {-74,252},{-74,-96},{60,-96},{60,-120}},   color={255,0,255}));
+  connect(abs2.y, greThr.u)
+    annotation (Line(points={{22,230},{38,230}}, color={0,0,127}));
+annotation (Documentation(info="<html>
+<p>
+This block implements a rule-based PID tuning method.
+</p>
+<p>
+The PID tuning method approximates the control process with a
+first-order time delayed (FOTD) model.
+It then determines the gain and delay of this FOTD model based on the responses of
+the control process to asymmetric relay feedback.
+After that, taking the gain and delay of this FOTD mode as inputs, this PID tuning
+method calculates the PID gains with an Approximate M-constrained Integral Gain Optimization (AMIGO) Tuner.
+</p>
+<p>
+This block is implemented using
+<a href=\"modelica://Buildings.Controls.OBC.Utilities.PIDWithInputGains\">
+Buildings.Controls.OBC.Utilities.PIDWithInputGains</a>
+and inherits most of its configuration. However, through the parameter
+<code>controllerType</code>, the controller can only be configured as PI or
+PID controller.
+</p>
+
+<h4>Autotuning Process</h4>
+<p>
+To use this block, place it in a control loop as any other PID controllers.
+Before the PID tuning process starts, this block is equivalent to
+<a href=\"modelica://Buildings.Controls.OBC.Utilities.PIDWithInputGains\">
+Buildings.Controls.OBC.Utilities.PIDWithInputGains</a>.
+This block starts the PID tuning process when a request for performing autotuning occurs, i.e.,
+when the value of the boolean input signal <code>triTun</code> changes from
+<code>false</code> to <code>true</code>.
+During the autotuning process, the output of the block changes into that of a relay controller
+(see <a href=\"modelica://Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.Controller\">
+Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.Controller</a>).
+The PID tuning process ends automatically
+(see details in
+<a href=\"modelica://Buildings.Controls.OBC.Utilities.PIDWithAutotuning.Relay.BaseClasses.TuningMonitor\">
+Buildings.Controls.OBC.Utilities.PIDWithAutotuning.BaseClasses.Relay.TunMonitor</a>),
+at which point this block turns back to a PID controller but with tuned PID parameters.
+</p>
+<p>
+<b>Note:</b> If an autotuning is ongoing, i.e., <code>inTunPro.y = true</code>,
+a new request for performing autotuning will be ignored.
+In addition, the set point should not be changed during the autotuning process.
```

Comment:
What will happen if the setpoint is changed during the tuning process? I see that you have a assert block`assMes3`. The assertion will give a warning not the error, so the tuning/simulation will not stop. So, should the tuning stop?

If the intention is to use the setpoint value at the moment when the tuning request becomes true, you may use the sampling value (`sam_u_s.y`) as input for the tuning process. If during the process the setpoint changes, it would give warning, but the tuning still use the old setpoint value. However, the question still is, should we allow the setpoint change?

Reply:
Yes, we could stop the tuning once the setpoint changes. However, I am not sure if changing setpoint is the only reason for the square root negative number. Thus, the current implementation stops the tuning once the square root negative number occurs.

Reply:
So I am assuming that when there is setpoint change during the tuning process, current implementation will stop the tuning process (no matter if or not the setpoint change causes the negative square root number)?

However, when check the result of `Buildings.Fluid.HeatExchangers.Examples.WetCoilCounterFlowPControlAutoTuning`, the variable `con.inTunPro.y` has following profile:
<img width="774" alt="Screenshot 2024-08-22 at 2 14 37 PM" src="https://github.com/user-attachments/assets/bc204ba5-c866-4825-bf16-1d5121f58d86">

It shows that although the setpoint `con.u_s` is changed at 2400 seconds, but the  `con.inTunPro.y` keeps being `true` till to 2493.84 seconds, meaning that the tuning process does not stop at 2400 seconds. Am I missing something?

Reply:
Sorry for the confusion.
In the current implementation,
- The tuning doesn't stop if the setpoint changes. A warning will be triggered.
- When the tuning stops, we will check if the tuning has been completed successfully or not. If not, we will use the control gains before the tuning.
- The flag that determines whether the tuning completes successfully or not will be false when the negative square root number occurs.

In my previous reply, I was saying we could change the current implementation into:
- The tuning stops if the setpoint changes. In this case, we will use the control gains before the tuning.

But my concern is that we cannot handle the case where the negative square root number is caused by other reasons.

Let me know if we need to have a quick call to discuss it.

Reply:
08/26/2024 meeting note:

-  Need to stop the tuning process once the setpoint changes. In this case, we will use the control gains before the tuning.

Reply:
Changes have been made in #3987  to stop the tuning process if the setpoint changes. In such a case, control gains will not be changed. If the setpoint is constant during the tuning, the tuning will be processed without interruption. We will check if the negative square root number occurs once the tuning is completed. If so, the tuning is considered to fail and control gains will not be changed.